In [1]:
# Importing necessary packages
import pandas as pd
import warnings
warnings.filterwarnings("ignore") #for the cleanliness of the notebook
import numpy as np

In [2]:
# Loading in the datasets
accounts_data = pd.read_csv('~sachamarois/Desktop/accounts.csv', sep=',')
user_data = pd.read_csv('~sachamarois/Desktop/user_usage_daily.csv', sep=',')

# Data Discovery

In [3]:
# Checking the number of rows and unique account ids  in the accounts table
print("The accounts table has",len(accounts_data), "rows with", accounts_data['account_id'].nunique(), "unique account ids")

The accounts table has 1513 rows with 1500 unique account ids


In [4]:
# Checking the number of rows in the users table
print("The users table has",len(user_data), "rows")

The users table has 929138 rows


In [5]:
# Quick overview of the statistical summary of the numerical columns in the accounts table
accounts_data.describe()

of_light_users  of_regular_users  of_admin_users           arr
count    1.513000e+03      1.513000e+03     1513.000000   1500.000000
mean     8.696259e+04      6.787654e+02     1329.872439  25971.200000
std      3.124514e+05      2.570905e+04    36345.317469  17543.103418
min      0.000000e+00      0.000000e+00        0.000000      0.000000
25%      0.000000e+00      0.000000e+00        0.000000      0.000000
50%      1.000000e+02      0.000000e+00        3.000000  30300.000000
75%      4.000000e+02      2.000000e+01        7.000000  41725.000000
max      3.999996e+06      1.000029e+06  1000000.000000  52900.000000

In [6]:
# Checking the data types in the accounts table
accounts_data.dtypes

name                 object
account_id           object
portal_id            object
type                 object
of_light_users      float64
of_regular_users    float64
of_admin_users      float64
company_size         object
churn_date           object
arr                 float64
dtype: object

In [7]:
# Checking the data types in the users table
user_data.dtypes

user_id       object
portal_id     object
queue_id      object
media_type    object
date          object
dtype: object

# Data Preparation

In [8]:
# Removing duplicates in the accounts table (that had rows as Ex-customer and Customer)
accounts_data = accounts_data[~accounts_data.duplicated(['account_id'], keep=False)]

# Converting the churn date object to a date type in the accounts table
accounts_data["churn_date"] = pd.to_datetime(accounts_data['churn_date'])

In [9]:
# Converting the date object to a date type in the users table
user_data['date'] = pd.to_datetime(user_data['date'])

In [10]:
# Merging the accounts table with the users table on portal_id
df = accounts_data.merge(user_data, left_on='portal_id', right_on='portal_id')

In [11]:
df

name  account_id        portal_id      type  \
0        Quisquam Ipsum Quiquia BV  ZF761HJR6E  AXOD97QQ1Q81HRU  Customer   
1        Quisquam Ipsum Quiquia BV  ZF761HJR6E  AXOD97QQ1Q81HRU  Customer   
2        Quisquam Ipsum Quiquia BV  ZF761HJR6E  AXOD97QQ1Q81HRU  Customer   
3        Quisquam Ipsum Quiquia BV  ZF761HJR6E  AXOD97QQ1Q81HRU  Customer   
4        Quisquam Ipsum Quiquia BV  ZF761HJR6E  AXOD97QQ1Q81HRU  Customer   
...                            ...         ...              ...       ...   
924005  Porro Dolore Etincidunt BV  MMXIXV6CKM  CKONRHHKYCQ2YGY  Customer   
924006    Ipsum Dolorem Dolore LLC  D7WDJV07IR  RA9PAY1T5G72MF4  Customer   
924007    Ipsum Dolorem Dolore LLC  D7WDJV07IR  RA9PAY1T5G72MF4  Customer   
924008    Ipsum Dolorem Dolore LLC  D7WDJV07IR  RA9PAY1T5G72MF4  Customer   
924009    Ipsum Dolorem Dolore LLC  D7WDJV07IR  RA9PAY1T5G72MF4  Customer   

        of_light_users  of_regular_users  of_admin_users  company_size  \
0                 50.0              10.0             2.0  3.Enterprise   
1                 50.0              10.0             2.0  3.Enterprise   
2                 50.0              10.0             2.0  3.Enterprise   
3                 50.0              10.0             2.0  3.Enterprise   
4                 50.0              10.0             2.0  3.Enterprise   
...                ...               ...             ...           ...   
924005           100.0              10.0             2.0  2.Mid-Market   
924006           100.0              10.0             2.0  2.Mid-Market   
924007           100.0              10.0             2.0  2.Mid-Market   
924008           100.0              10.0             2.0  2.Mid-Market   
924009           100.0              10.0             2.0  2.Mid-Market   

       churn_date      arr               user_id     queue_id media_type  \
0             NaT  44400.0  D2WAV77R38IS1RAW0M18  user.create        NaN   
1             NaT  44400.0  D2WAV77R38IS1RAW0M18   media.view      image   
2             NaT  44400.0  D2WAV77R38IS1RAW0M18   media.view      image   
3             NaT  44400.0  D2WAV77R38IS1RAW0M18  user.create        NaN   
4             NaT  44400.0  D2WAV77R38IS1RAW0M18   media.view      image   
...           ...      ...                   ...          ...        ...   
924005        NaT  33200.0  UHUZRQBZN4XWCCF8C15V   user.login        NaN   
924006        NaT  28700.0  D2WAV77R38IS1RAW0M18  user.create        NaN   
924007        NaT  28700.0  7O5ZQN09P3RJJLCCEH3C   user.login        NaN   
924008        NaT  28700.0  H2VSI0VZR5TVRMZQQT62   user.login        NaN   
924009        NaT  28700.0  SGKZ78A2WPA8NJJ1M572   user.login        NaN   

             date  
0      2021-06-16  
1      2021-10-04  
2      2021-09-11  
3      2021-10-28  
4      2021-10-02  
...           ...  
924005 2021-10-21  
924006 2021-10-27  
924007 2021-10-19  
924008 2021-11-29  
924009 2021-11-29  

[924010 rows x 14 columns]

# Question 1 

### Ex-customers

In [12]:
# Creating a new column 'days_until_churn' that calculates the number of days between the user event and the churn date
df['days_until_churn'] = (df['churn_date'] - df['date']).dt.days

In [13]:
# Creating an ex-customers data frame that includes only 'Ex-customers' and a positive 'days_until_churn',
# since we assume it is not possible to have usage after the churn date. This will also exclude the ex-customers
# without a churn date.

ex_customers = df[(df['type'] == 'Ex-customer') & (df['days_until_churn'] > 0)]

In [14]:
# Here we are counting the number of unique users grouped by account and type within the last 10 days of
# churing, a.k.a the number of engaged users within the last 10 days prior to churning.
ex_customers_use = ex_customers[ex_customers['days_until_churn'] <= 10]
ex_customers_use = ex_customers_use.groupby(['account_id','type'], as_index = False)['user_id'].agg('nunique')

In [15]:
# Calculating the average number of engaged users
print('The average number of engaged users within the last 10 days prior to churning, for ex-customers with activity, is',round(ex_customers_use['user_id'].sum()/len(ex_customers_use)),'.')

The average number of engaged users within the last 10 days prior to churning, for ex-customers with activity, is 7 .


This average number of engaged users only includes those ex-customers that had usage activity within the last 10 days of churning, but that does not give a global picture of the behavior of all customers that have churned. Hence why it makes sense to also include those that had no activity at all:

In [16]:
# Creating a list then a dataframe of those ex-customers that did not have any usage in the last 10 days prior 
# to churning and giving them a 0.
ex_customers_no_use = list(set(ex_customers_use['account_id']).symmetric_difference(set(ex_customers['account_id'])))
ex_customers_no_use = pd.DataFrame({'account_id':ex_customers_no_use,
                                    'user_id':0,
                                    'type':'Ex-customer'})

# Merging the two dataframes to get one table with all the ex customers and their usage
ex_customers_10days = pd.concat([ex_customers_use, ex_customers_no_use])

In [17]:
# Replacing the 'user_id' column name with a more intuitive column name 'no_engaged_users'
ex_customers_10days['no_engaged_users'] = ex_customers_10days['user_id']
ex_customers_10days = ex_customers_10days.drop('user_id', axis=1)

In [18]:
print('The average number of engaged users within the last 10 days prior to churning, for all ex-customers, is',round(ex_customers_10days['no_engaged_users'].sum()/len(ex_customers_10days)),'.')

The average number of engaged users within the last 10 days prior to churning, for all ex-customers, is 2 .


### Current customers

In [19]:
# Here, we are looking at the last 10 days of 2021 so we need to create a new column with the 31st of December.
df['end_2021'] = '31-12-2021'
df['end_2021'] = pd.to_datetime(df['end_2021'])

# Then we can calculate the difference between the last day of 2021 and the usage date
df['days_until_2022'] = (df['end_2021'] - df['date']).dt.days

In [20]:
# Creating a current customers data frame with the type 'Customers'
current_customers = df[df['type'] == 'Customer']

In [21]:
# Calculating the number of unique users grouped by account_id and type within the last 10 days of 2021
# a.k.a the number of engaged users per customer in the last 10 days of 2021
current_customers_use = current_customers[current_customers['days_until_2022'] <= 10]
current_customers_use = current_customers_use.groupby(['account_id','type'], as_index = False)['user_id'].agg('nunique')

In [22]:
# Calculating the average number of engaged users in the last 10 days for current customers that had usage activity.
print('The average number of engaged users within the last 10 days of 2021, for the current customers with activity, is',round(current_customers_use['user_id'].sum()/len(current_customers_use)),'.')

The average number of engaged users within the last 10 days of 2021, for the current customers with activity, is 19 .


Similarly to the ex-customers, this average number of engaged users only includes those current customers that had usage activity within the last 10 days of 2021, but that does not give a global picture of the behavior of all current customers. Hence why it makes sense to also include those that had no activity at all:

In [23]:
# Creating a list then a dataframe of those current customers that did not have any usage in the last 10 days of 2021
# and giving them a 0. 
current_customers_no_use = list(set(current_customers_use['account_id']).symmetric_difference(set(current_customers['account_id'])))
current_customers_no_use = pd.DataFrame({'account_id':current_customers_no_use,
                                         'user_id':0,
                                         'type':'Customer'})

# Merging the two dataframes to get one table with all the ex customers and their usage
current_customers_10days = pd.concat([current_customers_use, current_customers_no_use])

In [24]:
# Showing the difference between the number of current customers that had activity in the last 10 days vs no activity
print('The number of current customer accounts that had users in last 10 days is',len(current_customers_use),
     ', whereas the number of current customer accounts that did not have any usage is',len(current_customers_no_use),'.')

The number of current customer accounts that had users in last 10 days is 528 , whereas the number of current customer accounts that did not have any usage is 159 .


In [25]:
# Replacing the 'user_id' column name with a more intuitive column name 'no_engaged_users'
current_customers_10days['no_engaged_users'] = current_customers_10days['user_id']
current_customers_10days = current_customers_10days.drop('user_id', axis = 1)

In [26]:
print('The average number of engaged users within the last 10 days of 2021, across all curent customers, is',round(current_customers_10days['no_engaged_users'].sum()/len(current_customers_10days)),'.')

The average number of engaged users within the last 10 days of 2021, across all curent customers, is 15 .


It is quite clear from these results that there is a significant difference between the last 10 days of usage for the ex-customers (2) compared to the current customers (15). That means that current customers, on average, have almost 8 times more engaged users than the ex-customers. It is then safe to assume that a declining number or a low number of engaged users would indicate a path to churning.

### Engagement ratio of current customers

In [27]:
# Creating a column 'total_licenses' that calculates the total amount of licenses per current customer
current_customers['total_licenses'] = current_customers['of_light_users'] + current_customers['of_regular_users'] + current_customers['of_admin_users']
current_customers["total_licenses"] = current_customers['total_licenses'].astype(int)

In [28]:
# Calculating the number of unique users grouped by account_id and total_licenses
engagement_ratio_df = current_customers.groupby(['account_id','total_licenses'], as_index = False)['user_id'].agg('nunique')

# Renaming the user_id column to 'no_engaged_users' in order to have a clean data frame with the accounts,
# number of licenses and number of engaged users
engagement_ratio_df['no_engaged_users'] = engagement_ratio_df['user_id']
engagement_ratio_df = engagement_ratio_df.drop('user_id', axis = 1)
engagement_ratio_df

account_id  total_licenses  no_engaged_users
0    02JYV6YJOA              51                33
1    03ZB5GRO9X              62                68
2    050UI22KX7             276               117
3    07K5W1JL6L             592                 6
4    0982N2ATQY             218               112
..          ...             ...               ...
682  ZSB6ZWKHE9             726                 2
683  ZTBVHUAZUN               1               135
684  ZUJ2XORVOR             312               144
685  ZX9JPNBZUG               0               262
686  ZZRJO7VBHH             154                99

[687 rows x 3 columns]

In [29]:
# Calculating the engagement ratio by dividing the number of engaged users by the number of total licenses
engagement_ratio_df['engagement_ratio'] = engagement_ratio_df['no_engaged_users'] / engagement_ratio_df['total_licenses']
# Replacing the infinites, the ratios divided by 0, to nulls (Nan)
engagement_ratio_df['engagement_ratio'] = [np.nan if i == np.inf else i for i in engagement_ratio_df['engagement_ratio']]

In [30]:
print('The average engagement ratio for the current customers',engagement_ratio_df[engagement_ratio_df['engagement_ratio'] <= 1]['engagement_ratio'].mean(),'.')

The average engagement ratio for the current customers 0.31608820932210197 .


### Creating the CSV

In [31]:
# Creating a final current customers data frame with both the number of engaged users and the engagement ratio
current_customers_df = current_customers_10days.merge(engagement_ratio_df, left_on = 'account_id', right_on = 'account_id')
# Renaming the column to the correct 'no_engaged_users'
current_customers_df['no_engaged_users'] = current_customers_df['no_engaged_users_x']
current_customers_df = current_customers_df.drop(['no_engaged_users_x','no_engaged_users_y'], axis = 1)

current_customers_df

account_id      type  total_licenses  engagement_ratio  no_engaged_users
0    02JYV6YJOA  Customer              51          0.647059                 3
1    03ZB5GRO9X  Customer              62          1.096774                 4
2    050UI22KX7  Customer             276          0.423913                10
3    0982N2ATQY  Customer             218          0.513761                21
4    0DW1MF07OL  Customer            8009          0.143214                52
..          ...       ...             ...               ...               ...
682  PKIANOUOF0  Customer             115          0.052174                 0
683  3VCG8D2TN2  Customer             306          0.006536                 0
684  ECJUIS2GY9  Customer             611          0.003273                 0
685  P9LBSMWPQ8  Customer             112          0.017857                 0
686  IOGXPBWHVN  Customer             100          0.430000                 0

[687 rows x 5 columns]

In [32]:
# Creating a final ex-customers dataframe with an extra column 'engagement ratio' with all nulls to be able to merge 
# with the current customer df
ex_customers_df = ex_customers_10days
ex_customers_df['engagement_ratio'] = np.nan
ex_customers_df

account_id         type  no_engaged_users  engagement_ratio
0   3C1ZA0DV6N  Ex-customer                 9               NaN
1   3KX7EEDTJV  Ex-customer                 2               NaN
2   A27F8LWVV5  Ex-customer                 1               NaN
3   CZUCQMRUNJ  Ex-customer                 6               NaN
4   DJT3BR4F2Z  Ex-customer                22               NaN
5   GE0Q5ZJAUC  Ex-customer                 6               NaN
6   HLX0ST69WW  Ex-customer                 2               NaN
7   UNLPAIH0P4  Ex-customer                14               NaN
8   XPVMSR75RM  Ex-customer                 6               NaN
9   XVHAGKC4LO  Ex-customer                 2               NaN
10  Z0HR9T9CD1  Ex-customer                 2               NaN
0   C0TR58PS0R  Ex-customer                 0               NaN
1   GGXP5CITUF  Ex-customer                 0               NaN
2   K5J42OI6OA  Ex-customer                 0               NaN
3   2HXRJM4T8B  Ex-customer                 0               NaN
4   YHONBPQGA6  Ex-customer                 0               NaN
5   G77Q9VJUHB  Ex-customer                 0               NaN
6   S6R6BILQPB  Ex-customer                 0               NaN
7   39PECN0DL5  Ex-customer                 0               NaN
8   61B7PXJBC8  Ex-customer                 0               NaN
9   HEJLA0GNZS  Ex-customer                 0               NaN
10  EGDD8MWE77  Ex-customer                 0               NaN
11  T1NTMJSRD9  Ex-customer                 0               NaN
12  7Q0I9BL14H  Ex-customer                 0               NaN
13  B34KUQAAJ8  Ex-customer                 0               NaN
14  5ELTFEBE54  Ex-customer                 0               NaN
15  09LZRK79UF  Ex-customer                 0               NaN
16  WIXZ2C9ROS  Ex-customer                 0               NaN
17  F60YKNFU84  Ex-customer                 0               NaN

In [33]:
# Creating the final dataframe with both ex-customers and current-customers
final_df = pd.concat([current_customers_df, ex_customers_df])
final_df

account_id         type  total_licenses  engagement_ratio  \
0   02JYV6YJOA     Customer            51.0          0.647059   
1   03ZB5GRO9X     Customer            62.0          1.096774   
2   050UI22KX7     Customer           276.0          0.423913   
3   0982N2ATQY     Customer           218.0          0.513761   
4   0DW1MF07OL     Customer          8009.0          0.143214   
..         ...          ...             ...               ...   
13  B34KUQAAJ8  Ex-customer             NaN               NaN   
14  5ELTFEBE54  Ex-customer             NaN               NaN   
15  09LZRK79UF  Ex-customer             NaN               NaN   
16  WIXZ2C9ROS  Ex-customer             NaN               NaN   
17  F60YKNFU84  Ex-customer             NaN               NaN   

    no_engaged_users  
0                  3  
1                  4  
2                 10  
3                 21  
4                 52  
..               ...  
13                 0  
14                 0  
15                 0  
16                 0  
17                 0  

[716 rows x 5 columns]

In [34]:
# Exporting the CSV
final_df.to_csv('final_df.csv')

# Question 2

#### 2a
From the findings above, it is clear that customers with a declining amount of engaged users or a very low amount of engaged users is an indication towards churn. In this case, the company should consistently track the number of engaged users on a individual customer basis to highlight any downward trends or low tresholds. This could be visualised through a dashboard for the customer success or account management team for instance. Once these customers have been flagged, there are a few options possible: 
- Direct communication with the customer to understand why there is a declining or low amount of engaged users, would be helpful to find a solution to increase that engagement and show the benefits the customer would get from an uptake in engagment. 
- Another solution could be to offer more training to the users of a customer in order to increase the engagement and show the value of the product to those users.
- Finally, the company could also use a model that predicts the likelihood of customers churning based on their engagement ratio and target those that are most likely to churn by engaging with them through "check-ins" or exciting new features that have been released, or are "soon-to-be" released.  

#### 2b
Once business objectives have been reviewed and the KPI is clearly defined and agreed upon across the company, it is possible to review the data and decide on an target for that KPI. In this case, from the analysis above, the average engagement ratio is approximately 31.6%. Let's assume the objective of the company is to retain customers longer by increasing this engagment ratio. A helathy increase here would be around 5-10% over the year 2022. 

To make this actionable, we could also set a short term target of an increase of 3% by the end of Q2 so we can potentially end up in the 5 to 10 percent range by the end of the year. 
To summarise, setting a target of 35% engagement ratio by the end of Q2 2022 and a 39% engagement ratio by the end of 2022, would allow a review of our KPI throughout the year and assess the situation.